### ISSUE TITLE
Understanding RTT semantics

### ISSUE THREAD

**Issue by kripken** [WebAssembly/gc#170]
Probably just me (not having a CS background) but personally I've found this a little hard to follow. Perhaps some more examples or reftests might help - or are those present somewhere I didn't find?

One specific question is what RTT casts are valid. For example,
```wat
x = (rtt.canon A)
y = (rtt.sub B x)
z = (rtt.sub C y)

m = (rtt.canon A)
n = (rtt.sub C m)
```
Here `z -> y -> x`, and `n -> m`, with similar types except the second chain "skips" the type B in the middle. Is it valid to cast an object with RTT `z` using an RTT `n`? The object has the right wasm type (C in both cases), but different RTT "chains". So it's memory safe, but perhaps not valid to cast?

**Comment by conrad-watt** [comment]
In the hope of invoking Cunningham's Law, my understanding is that the cast would be valid.

This is based on the definition of `rtt.canon/sub` [here](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#runtime-types-1), `ref.cast` [here](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#casts) and sub-RTT [here](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#runtime-types).

In particular, if I have an an object O of static type `A` and RTT `z`  and I call `rtt.cast C O n`, the cast is specified as succeeding if `z` is a sub-RTT of `n`. The RTTs `z` and `n` represent the same static type, so the cast should succeed.

I also believe that the cast should succeed even if `n` is merely the RTT of an _equivalent_ type to `C`, in the sense defined [here](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md#type-equivalence). This is tied up with the expectation that all types are canonicalised, so that equivalent types are collapsed into each other, and `rtt.canon/sub` under the hood key off the canonicalised "type equivalence class" rather than just the type index. See the algorithms @rossberg references [here](https://github.com/WebAssembly/gc/issues/68).

**Comment by RossTate** [comment]
According to the given definition of `rtt.sub`, the cast succeeds if and only if `B` and `C` are dynamically equivalent, i.e. equivalent after imported types are instantiated. The reason is that distinct executions of `rtt.sub` produce the some result if and only if the input rtt is the same and the types are dynamically equivalent. So `x` and `m` are guaranteed to be the same, but `y` and `n` are the same if and only if `B` and `C` are dynamically equivalent.

**Comment by jakobkummerow** [comment]
What Ross said. An Object with RTT `z` can be `ref.cast` to `x`, `y`, or `z`, so the cast only succeeds if `n` happens to be pointer-identity-equal to `y`.

Note: the current implementation of RTTs in the V8 prototype does not consider structural type equivalence, so it would always produce distinct RTTs for `y` and `n` whenever `B` and `C` are distinct entries in the module's type section.

One way to think about it: RTT inheritance (as created by `rtt.sub`) is nominal, so explicitly more restrictive than what memory safety would allow. One underlying motivation is to make runtime type checks fast: depending on implementation choices, at worst linear in the length of the `z→y→x` chain, at best constant-time by making use of the RTT depth value and comparing exactly one candidate.

**Comment by conrad-watt** [comment]
In this case the definition of [sub-RTT](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#runtime-types) in the MVP needs to be changed, because it is true that the RTTs `z` and `n` represent the same static type, and hence are sub-RTTs of each other. The `ref.cast` operation is only [described](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#casts) in terms of this sub-RTT check.

**Comment by conrad-watt** [comment]
But now I think about it, this has some strange consequences. For example,
```
x = (rtt.canon A)
y = (rtt.sub B x)

m = (rtt.canon A)
n = (rtt.sub B n)
```
My interpretation of [this](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#runtime-types-1) is that it's still required that `y` and `n` have pointer (physical) equality. So in a full implementation, RTTs must be memoized?

In my head RTTs were each a vector of (pointers to) canonicalised types, and `rtt.sub T parent` copied/referenced the `parent` RTT vector and extended it with the canonicalised T. If downcasting were done by directly comparing canonicalised type pointers, there would be no need to explicitly memoize the results of `rtt.sub`. But clearly my mental model is wrong given the semantics above.

**Comment by RossTate** [comment]
Your mental model is reasonable. Remember that this proposal was originally designed to use structural casts rather than nominal casts, which explains the inconsistency in its model. Another reflection of that is the fact that the JS API is not stable under type equivalence, meaning replacing a type with a (statically) equivalent type in an import/export's signature changes the semantics of the API.

**Comment by conrad-watt** [comment]
Right now there's certainly an inconsistency within the [runtime types](url) subsection of the proposal. Either the note about implementation strategy needs to change (blessing my mental model), or the definition of sub-RTT needs to change (blessing @jakobkummerow's expected semantics).

Assuming type canonicalisation needs to be implemented either way, I believe my model is less complicated and no slower than the alternative. It would make the `rtt` argument to `ref.cast` somewhat redundant, which is a point against it being the "intended semantics". That being said, the apparent alternative requirement to memoize `rtt.sub` results on top of type canonicalisation seems very weird, and relaxing this in isolation seems to get us close to `rtt.fresh`.

We'll need @rossberg to weigh in on whether there's an intended behaviour that's already been thought through, and edit the proposal accordingly.

**Comment by jakobkummerow** [comment]
I think the inconsistency is resolved if you read the part "r1 and r2 represent the same static type" with the implied notion that the "static type" of an RTT includes its depth. (+1 to clarifying that in the spec text!) So `(rtt 2 $C)` and `(rtt 3 $C)` are *not* the same static RTT type.

I think your recent example has a typo and you mean `n = (rtt.sub B m)`; aside from that I agree with it. Both `rtt.canon` and `rtt.sub` can be invoked repeatedly with the same arguments and will yield the same results:
```
x1 = (rtt.canon A)
x2 = (rtt.canon A)
y1 = (rtt.sub B x1)
y2 = (rtt.sub B x1)
y3 = (rtt.sub B x2)
```
In that case, `x1` and `x2` are pointer-identical, so are `y1` and `y2`, and because `x1 == x2` we also get pointer-equality between `y2` and `y3`.

(For clarity's sake: that example would work as described in our prototype; the only thing we haven't implemented is to make sure that the same canonicalizing behavior holds for structurally equivalent types `A'` and `B'` being used in place of `A` and `B`.)

**Comment by conrad-watt** [comment]
> I think the inconsistency is resolved if you read the part "r1 and r2 represent the same static type" with the implied notion that the "static type" of an RTT includes its depth. (+1 to clarifying that in the spec text!) So `(rtt 2 $C)` and `(rtt 3 $C)` are not the same static RTT type.

I don't think this completely captures the issue. For example, consider this:

```
x = (rtt.canon A)
y = (rtt.sub P x)
z = (rtt.sub C y)

m = (rtt.canon A)
n = (rtt.sub Q m)
o = (rtt.sub C n)
```

Under the semantics you laid out, assuming the results of two calls to `rtt.sub C` are pointer-equivalent iff their inputs are pointer-equivalent, using `o` to cast an object with RTT `z` must `trap` if `P !== Q`.

However even assuming a depth-based definition of "static type", `z` and `o` would be sub-RTTs.

> I think your recent example has a typo 

Oops, fixed

> For clarity's sake: that example would work as described in our prototype;

Does this mean that V8 is currently memoizing `rtt.sub` results?

**Comment by RossTate** [comment]
@conrad-watt does have a consistent model and is right that the spec needs to be clarified. These issues also come up with the parameterized extension to `rtt.canon/sub` in the Post-MVP, where it is similarly unclear if the specific input rtts matter, and it is is furthermore unclear how issues like variance are supposed to be accommodated. At present I cannot figure out how casting for C#/Java's covariant reference arrays or Kotlin's invariant-but-projectable native arrays are supposed to be implemented with this casting approach (presuming that they are supposed to use `rtt.canon/sub` due to being primitive in the respective languages).

**Comment by jakobkummerow** [comment]
The P/Q example is insightful. I agree that that scenario is not adequately covered by the current spec text. If I understand the spec's intention correctly, then an object with RTT `z` should not be castable to `o` (or `n`), but the current wording around static types doesn't express that.

> Does this mean that V8 is currently memoizing `rtt.sub` results?

Yes, in the sense that repeated invocations of `rtt.sub` with the same arguments (parent RTT + static target type) will return the same (i.e. pointer-identical) result.
The spec deviation we currently have in the prototype is that if the static target type is structurally equivalent but a distinct entry in the types section, then we don't canonicalize that (and instead return a different RTT). If the RTT system remains what it currently is, then we will of course have to fix that eventually (by implementing the whole automaton-minimization algorithm).

**Comment by rossberg** [comment]
Sorry for the delay.

Yes, indeed the def of sub-RTT in the MVP doc is missing something. Thanks for pointing out! The intention is what @jakobkummerow said. More precisely, this case in the definition:

>   - r1 and r2 represent the same static type

should be

>   - r1 and r2 represent the same static type and have equal parents

So to answer @kripken's OP, given `C <: B1 < A` and `C <: B2 <: A` statically, but `B1 =/= B2`, and
```
a = (rtt.canon A)

b1 = (rtt.sub B1 a)
c1 = (rtt.sub C b1)

b1' = (rtt.sub B1 a)
c1' = (rtt.sub C b1')

b2 = (rtt.sub B2 a)
c2 = (rtt.sub C b2)

c3 = (rtt.sub C a)
```
we have `b1 = b1'` and `c1 = c1'`, but `c1`, `c2`, and `c3` are all different.

The general idea is that the sub-RTT relation is a sub relation of the static one that is always tree-shaped, in order to allow efficient runtime tests. 

Plese see #177 for a fix.

**Comment by conrad-watt** [comment]
@rossberg, was the semantics I sketched out [above](https://github.com/WebAssembly/gc/issues/170#issuecomment-748231218) ever considered? MVP runtime tests would still be constant time, but does it have post-MVP implications?

With the current semantics, aside from the extra memoization complication in `rtt.sub`, it's harder for modules to abstract away supertypes in the hierarchy that aren't relevant. Imagine I have types `C <: B <: A`, and modules `M` and `N` such that `M` only manipulates objects of types `C` and `B`, but `N` uses objects of all three types. I might want to create an `C` object in `M`, and pass it to `N`  with a static type of `B`, and attempt in `N` to downcast that object to static type `C`. Currently for the downcast to work, `M` needs to either import `N`'s rtts, or include a definition of `A` that it can call `rtt.canon` on. Otherwise when `M` creates a `C` object, the rtt of the object will be incompatible with `N`'s rtts.

**Comment by rossberg** [comment]
@conrad-watt, I'm not sure I understand how your semantics differs. The way you describe, it sounds equivalent? Or maybe I misunderstand what exactly you mean by "directly comparing canonicalised type pointers".

The essential property you want is that the subtype check is constant-time. I'm not sure how you would achieve that while being agnostic to the depth of the subtype chain. Can you elaborate?

**Comment by conrad-watt** [comment]
In the current semantics, AFAIU the efficient representation of an RTT is as a fixed-length vector of pointers to parent RTTs.

In my sketched semantics, an RTT is a fixed-length vector of (pointers to) canonicalised types. For example, let's say for the sake of argument that there's a synchonized datastructure in the runtime which holds a single entry for each canonicalised type. Then an RTT could be represented as a fixed-length vector of pointers to entries in that datastructure. Alternatively, you could give each canonicalised type an enumerated colour and then each RTT would be a fixed-length vector of colours.

You're right though, even with my approach you can't actually mix objects with different subtype chain lengths and expect constant time behaviour. So my second point is wrong, the only thing you avoid with my idea is the `rtt.sub` memoization. That is, instead of `rtt.sub` with the same inputs returning RTTs which are physically equal, instead you get RTTs with the same canonicalised type vectors.

**Comment by rossberg** [comment]
I'm still not sure I understand. IIUC, this would not create a (tree-shaped) type hierarchy, but what does it create? A cast may "accidentally" succeed across unrelated hierarchies when they happen to have the same type at the same depth. That may not necessarily be a practical problem, but seems a bit arbitrary. Semantically, it certainly amounts to a more complicated and ad-hoc definition of sub-RTT that essentially prescribes a very specific implementation technique.

Furthermore, how would it work for casts between imported types (or over polymorphic types, under the polymorphism extension)? In that case, you do not generally know the index. The proposal encapsulates that away in the RTT, or allows other implementation techniques entirely. It optionally allows to expose the depth statically for potential optimisation reasons where known, but does not depend on that (and that feature may be a premature optimisation that we might want to drop, as discussed somewhere else).

There is another issue with the signature you suggest, although I think that is an independent choice. It's a central point of the design that RTT values are explicit operands. That's in order to stay forward-compatible with the introduction parametric(!) polymorphism. Without it, you'd have to build in all the complexity and cost of reified generics into core Wasm's semantics and its calling conventions for polymorphism everywhere.

**Comment by conrad-watt** [comment]
I think we can shortcut a deep dive into this. I'm happy with the current model and I don't want to give the impression that I'm going to push for any alternative.
